File: src/shared/roadtrip/hooks/scrub-paint.ts
/**
 * Painting the scrub — one frame of «&nbsp;Défilé&nbsp;», read off the plan.
 *
 * Everything drawn here is a reading of `ScrubPlan` at `t`, so the tape, the
 * flash and the shutter dip agree with each other and with the numeral the
 * badge steps above them. Painted between the picture and the shades (the
 * engine's seam), so the badge always sits on top.
 *
 * What a frame shows, while the sweep runs:
 *
 * - a stop with a picture → that picture, cover-cropped into the frame (the
 *   source picture of the piece telling that day, or one the author picked);
 * - a stop without one, or whose picture could not be found → the frame goes
 *   dark. It is not a stand-in and it is not the hero shown early: an empty
 *   day looks empty, which is the honest reading of the calendar and what
 *   makes the hero's arrival land.
 *
 * Once the sweep has come to rest this draws nothing but the tape, and the
 * piece's own picture — already on the canvas — is the frame.
 *
 * Sizes are in units of a 1080-wide frame, so the stage preview and a 1080×1920
 * export draw the same tape at two scales.
 */

import { drawFramed } from '../../media/framing';
import type { FrameBox, HookCtx2D, HookPicture } from './hook-variant';
import {
  edgeFadeAt,
  hexToRgba,
  tapeFraction,
  tapeGeometry,
  tapeTicks,
  type ScrubOptions,
  type ScrubPlan,
} from './scrub-plan';

/** What an untold day looks like: the paper's ink, not pure black. */
const EMPTY_DAY = '#0c0b09';
/** How long the shutter dip lasts after the head lands, in seconds. */
const DIP_SECONDS = 0.06;
/** How dark the dip gets at its deepest. */
const DIP_ALPHA = 0.38;

export function paintScrub(
  g: HookCtx2D,
  plan: ScrubPlan,
  opts: ScrubOptions,
  pictures: ReadonlyMap<string, HookPicture> | undefined,
  t: number,
  frame: FrameBox,
): void {
  const { width: w, height: h } = frame;
  if (w <= 0 || h <= 0) return;

  const sweeping = plan.sweepSeconds > 0 && t < plan.endSeconds;
  const index = plan.stopAt(t);

  if (opts.flash && sweeping) {
    const stop = plan.stops[index];
    const picture = stop.pictureKey ? pictures?.get(stop.pictureKey) : undefined;
    if (!drawPicture(g, picture, w, h)) {
      g.fillStyle = EMPTY_DAY;
      g.fillRect(0, 0, w, h);
    }
    // The projector's advance: a frame of darkness as the head lands. Not on
    // the first stop, which is where the sweep starts rather than a landing.
    const since = plan.sinceStopAt(t);
    if (index > 0 && since >= 0 && since < DIP_SECONDS) {
      g.save();
      g.globalAlpha = DIP_ALPHA * (1 - since / DIP_SECONDS);
      g.fillStyle = '#000';
      g.fillRect(0, 0, w, h);
      g.restore();
    }
  }

  paintTape(g, plan, opts, t, w, h);
}

/**
 * Cover-crop a day's picture into the frame; false when there is none to
 * draw. A bitmap closed under a render still in flight throws on `drawImage` —
 * that frame then shows the day as empty rather than killing the paint loop.
 */
function drawPicture(
  g: HookCtx2D,
  picture: HookPicture | undefined,
  w: number,
  h: number,
): boolean {
  if (!picture || picture.width <= 0 || picture.height <= 0) return false;
  try {
    drawFramed(g, picture.image, picture.width, picture.height, w, h);
    return true;
  } catch {
    return false;
  }
}

function paintTape(
  g: HookCtx2D,
  plan: ScrubPlan,
  opts: ScrubOptions,
  t: number,
  w: number,
  h: number,
): void {
  // Every size and the band come from the pure geometry, so the rectangle the
  // stage grabs (`tapeBox`) is the one drawn here.
  const { x0, x1, length, baseline, dir, u, shortTick, headTall, tallTick, band } = tapeGeometry(w, h, opts);
  const headDay = plan.headDayAt(t);
  const legs = new Set(plan.legStarts);
  const xOf = (day: number) => x0 + length * tapeFraction(day, plan.totalDays);
  const fade = (x: number) => edgeFadeAt(x, x0, x1, opts.edgeFade);

  g.save();

  if (opts.tapeBackground) {
    // A dark band, rounded, faded at the ends with everything else when the
    // fade is on — a gradient fill rather than a mask, so no compositing mode
    // is involved (a shadow is dropped under `destination-*`, `studio.md`).
    const { x: bx, y: by, width: bw, height: bh } = band;
    if (opts.edgeFade) {
      const grad = g.createLinearGradient(bx, 0, bx + bw, 0);
      const stops = 12;
      for (let i = 0; i <= stops; i++) {
        const k = i / stops;
        grad.addColorStop(k, `rgba(0,0,0,${(opts.backgroundOpacity * fade(bx + bw * k)).toFixed(3)})`);
      }
      g.fillStyle = grad;
    } else {
      g.fillStyle = `rgba(0,0,0,${opts.backgroundOpacity})`;
    }
    roundedBar(g, bx, by, bw, bh, 8 * u);
  }

  if (opts.showTrack) {
    if (opts.edgeFade) {
      const grad = g.createLinearGradient(x0, 0, x1, 0);
      const stops = 12;
      for (let i = 0; i <= stops; i++) {
        const k = i / stops;
        grad.addColorStop(k, hexToRgba(opts.tickColor, opts.tickOpacity * fade(x0 + length * k)));
      }
      g.fillStyle = grad;
    } else {
      g.fillStyle = hexToRgba(opts.tickColor, opts.tickOpacity);
    }
    g.fillRect(x0, baseline - 0.75 * u, length, 1.5 * u);
  }

  for (const day of tapeTicks(plan.totalDays, length, plan.legStarts, opts.tickGap * u)) {
    const leg = legs.has(day);
    const tall = leg ? tallTick : shortTick;
    const passed = day <= headDay + 1e-6;
    const x = xOf(day);
    const alpha = fade(x) * (passed ? 0.95 : Math.min(1, opts.tickOpacity + (leg ? 0.25 : 0)));
    if (alpha <= 0) continue;
    g.fillStyle = hexToRgba(passed ? opts.passedColor : opts.tickColor, alpha);
    g.fillRect(x - u, dir < 0 ? baseline - tall : baseline, 2 * u, tall);
  }

  // The reading head. Its glow is one shadow blur a frame — affordable on a
  // single small shape where it is not on every tick.
  const hx = x0 + length * tapeFraction(headDay, plan.totalDays);
  g.globalAlpha = fade(hx);
  if (opts.headGlow) {
    g.shadowColor = hexToRgba(opts.passedColor, 0.75);
    g.shadowBlur = 14 * u;
  }
  g.fillStyle = hexToRgba(opts.passedColor, 1);
  if (opts.headStyle === 'dot') {
    const cy = dir < 0 ? baseline - shortTick - 9 * u : baseline + shortTick + 9 * u;
    g.beginPath();
    g.arc(hx, cy, 7 * u, 0, Math.PI * 2);
    g.fill();
  } else if (opts.headStyle === 'needle') {
    // A triangle pointing AT the tape from beyond the tallest tick.
    const base = dir < 0 ? baseline - headTall : baseline + headTall;
    const tip = dir < 0 ? baseline - 4 * u : baseline + 4 * u;
    g.beginPath();
    g.moveTo(hx, tip);
    g.lineTo(hx - 7 * u, base);
    g.lineTo(hx + 7 * u, base);
    g.closePath();
    g.fill();
  } else {
    roundedBar(
      g,
      hx - 2.5 * u,
      dir < 0 ? baseline - headTall + 6 * u : baseline - 6 * u,
      5 * u,
      headTall,
      2.5 * u,
    );
  }

  g.restore();
}

function roundedBar(g: HookCtx2D, x: number, y: number, w: number, h: number, r: number): void {
  g.beginPath();
  g.moveTo(x + r, y);
  g.lineTo(x + w - r, y);
  g.quadraticCurveTo(x + w, y, x + w, y + r);
  g.lineTo(x + w, y + h - r);
  g.quadraticCurveTo(x + w, y + h, x + w - r, y + h);
  g.lineTo(x + r, y + h);
  g.quadraticCurveTo(x, y + h, x, y + h - r);
  g.lineTo(x, y + r);
  g.quadraticCurveTo(x, y, x + r, y);
  g.closePath();
  g.fill();
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Roadtrip/Hooks/ScrubPaint.swift
// One frame of «Défilé», read off the plan — the layout arithmetic of
// `src/shared/roadtrip/hooks/scrub-paint.ts`. The strokes themselves (a fill,
// a rounded bar, a gradient, a shadow) are the app's Core Graphics painter;
// every position, size, colour and alpha it draws is decided here, so the
// stage and the export draw the same tape at two scales and the rectangle a
// click grabs (`tapeBox`) is the band that is drawn.
//
// What a frame shows while the sweep runs:
// - a stop with a picture → that picture, cover-cropped into the frame (the
//   source picture of the piece telling that day, or one the author picked);
// - a stop without one, or whose picture could not be found → the frame goes
//   dark (`scrubEmptyDayColor`). Not a stand-in and not the hero shown early:
//   an empty day looks empty, which is what makes the hero's arrival land;
// - a frame of darkness as the head lands on any stop but the first — the
//   projector's advance, `scrubDipSeconds` long.
// Once the sweep has come to rest nothing but the tape is drawn, and the
// piece's own picture — already on the canvas — is the frame.
//
// Sizes are in units of a 1080-wide frame (`TapeGeometry.u`).

import Foundation

/// What an untold day looks like: the paper's ink, not pure black.
public let scrubEmptyDayColor = "#0c0b09"
/// How long the shutter dip lasts after the head lands, in seconds.
public let scrubDipSeconds = 0.06
/// How dark the dip gets at its deepest.
public let scrubDipAlpha = 0.38
/// How many steps a faded band or track's gradient is drawn with (13 stops).
public let scrubFadeSteps = 12

/// What fills the frame while the head sweeps.
public enum ScrubFlash: Equatable, Sendable {
    /// The picture decoded under this key (`hookPictureKey`), cover-cropped
    /// into the frame. Should the app fail to draw it after all (an image
    /// released under a render still in flight), it paints the empty day's
    /// ink, exactly as the web falls back.
    case picture(key: String)
    /// A stop with nothing to show.
    case empty
}

/// One stop of a horizontal alpha gradient, `offset` 0…1 along the bar.
public struct ScrubGradientStop: Equatable, Sendable {
    public var offset: Double
    public var alpha: Double
}

/// A filled bar of the tape — the band behind it or the track under it.
public struct ScrubBar: Equatable, Sendable {
    public var rect: Rect
    /// Corner radius; 0 is square.
    public var radius: Double
    /// `#rrggbb`.
    public var color: String
    /// The fill's alpha when it is flat.
    public var alpha: Double
    /// With the edge fade on: the alpha along the bar, left to right, in place of `alpha`.
    public var gradient: [ScrubGradientStop]?
}

/// One tick of the tape.
public struct ScrubTick: Equatable, Sendable {
    public var day: Int
    public var rect: Rect
    /// `#rrggbb` — the passed colour once the head has gone by.
    public var color: String
    public var alpha: Double
    /// A leg starts on this day: the tall tick.
    public var leg: Bool
    public var passed: Bool
}

/// The reading head's shape, in frame pixels.
public enum ScrubHeadShape: Equatable, Sendable {
    /// A rounded upright bar.
    case bar(Rect, radius: Double)
    /// A disc above (or below) the ticks.
    case dot(center: Point, radius: Double)
    /// A triangle pointing AT the tape from beyond the tallest tick.
    case needle(tip: Point, left: Point, right: Point)
}

/// The head as drawn: its shape, its colour and alpha, and its glow.
public struct ScrubHeadMark: Equatable, Sendable {
    public var shape: ScrubHeadShape
    /// `#rrggbb`, drawn at full opacity under `alpha`.
    public var color: String
    /// The whole head's alpha — the edge fade at its position.
    public var alpha: Double
    /// The glow's blur radius (one shadow a frame) in the passed colour at
    /// 0.75, or nil with the glow off.
    public var glowBlur: Double?
}

/// Everything one frame of Défilé draws, in painting order: the flash and its
/// dip (under the tape), then the band, the track, the ticks and the head.
public struct ScrubFrame: Equatable, Sendable {
    /// What fills the frame, or nil to leave the picture below as it is.
    public var flash: ScrubFlash?
    /// Black over the flash at this alpha; 0 for none.
    public var dipAlpha: Double
    public var band: ScrubBar?
    public var track: ScrubBar?
    public var ticks: [ScrubTick]
    public var head: ScrubHeadMark
}

/// Round to three decimals, as the web's `toFixed(3)` alpha does.
private func scrubAlpha3(_ v: Double) -> Double {
    (v * 1000).rounded() / 1000
}

/// One frame of the sweep at `t` on a frame of `frame`'s size, or nil for a
/// frame with no area (the web draws nothing). A stop flashes its picture only
/// when `pictures` holds one with an area under its key; otherwise it is dark.
public func scrubFrame(_ plan: ScrubPlan, _ opts: ScrubOptions, _ pictures: [String: HookPicture]?,
                       _ t: Double, _ frame: FrameBox) -> ScrubFrame? {
    let w = frame.width
    let h = frame.height
    guard w > 0, h > 0 else { return nil }

    let sweeping = plan.sweepSeconds > 0 && t < plan.endSeconds
    let index = plan.stopAt(t)

    var flash: ScrubFlash? = nil
    var dipAlpha = 0.0
    if opts.flash && sweeping {
        let key = plan.stops[index].pictureKey
        let picture = key.flatMap { pictures?[$0] }
        if let key, let picture, picture.width > 0, picture.height > 0 {
            flash = .picture(key: key)
        } else {
            flash = .empty
        }
        // The projector's advance: not on the first stop, which is where the
        // sweep starts rather than a landing.
        let since = plan.sinceStopAt(t)
        if index > 0 && since >= 0 && since < scrubDipSeconds {
            dipAlpha = scrubDipAlpha * (1 - since / scrubDipSeconds)
        }
    }

    let g = tapeGeometry(w, h, opts.tapeGeometryOptions)
    let u = g.u
    let headDay = plan.headDayAt(t)
    let legs = Set(plan.legStarts)
    func xOf(_ day: Double) -> Double { g.x0 + g.length * tapeFraction(day, plan.totalDays) }
    func fade(_ x: Double) -> Double { edgeFadeAt(x, g.x0, g.x1, opts.edgeFade) }
    let steps = scrubFadeSteps

    var band: ScrubBar? = nil
    if opts.tapeBackground {
        let b = g.band
        let gradient: [ScrubGradientStop]? = opts.edgeFade
            ? (0...steps).map { i in
                let k = Double(i) / Double(steps)
                return ScrubGradientStop(offset: k, alpha: scrubAlpha3(opts.backgroundOpacity * fade(b.x + b.width * k)))
            }
            : nil
        band = ScrubBar(rect: b, radius: 8 * u, color: "#000000", alpha: opts.backgroundOpacity, gradient: gradient)
    }

    var track: ScrubBar? = nil
    if opts.showTrack {
        let gradient: [ScrubGradientStop]? = opts.edgeFade
            ? (0...steps).map { i in
                let k = Double(i) / Double(steps)
                return ScrubGradientStop(offset: k, alpha: opts.tickOpacity * fade(g.x0 + g.length * k))
            }
            : nil
        track = ScrubBar(rect: Rect(x: g.x0, y: g.baseline - 0.75 * u, width: g.length, height: 1.5 * u),
                         radius: 0, color: opts.tickColor, alpha: opts.tickOpacity, gradient: gradient)
    }

    var ticks: [ScrubTick] = []
    for day in tapeTicks(plan.totalDays, g.length, plan.legStarts, minGapPx: opts.tickGap * u) {
        let leg = legs.contains(day)
        let tall = leg ? g.tallTick : g.shortTick
        let passed = Double(day) <= headDay + 1e-6
        let x = xOf(Double(day))
        let ahead = min(1, opts.tickOpacity + (leg ? 0.25 : 0))
        let alpha = fade(x) * (passed ? 0.95 : ahead)
        if alpha <= 0 { continue }
        let y = g.dir < 0 ? g.baseline - tall : g.baseline
        ticks.append(ScrubTick(day: day, rect: Rect(x: x - u, y: y, width: 2 * u, height: tall),
                               color: passed ? opts.passedColor : opts.tickColor, alpha: alpha,
                               leg: leg, passed: passed))
    }

    let hx = xOf(headDay)
    let shape: ScrubHeadShape
    switch opts.headStyle {
    case .dot:
        let cy = g.dir < 0 ? g.baseline - g.shortTick - 9 * u : g.baseline + g.shortTick + 9 * u
        shape = .dot(center: Point(hx, cy), radius: 7 * u)
    case .needle:
        let base = g.dir < 0 ? g.baseline - g.headTall : g.baseline + g.headTall
        let tip = g.dir < 0 ? g.baseline - 4 * u : g.baseline + 4 * u
        shape = .needle(tip: Point(hx, tip), left: Point(hx - 7 * u, base), right: Point(hx + 7 * u, base))
    case .bar:
        let y = g.dir < 0 ? g.baseline - g.headTall + 6 * u : g.baseline - 6 * u
        shape = .bar(Rect(x: hx - 2.5 * u, y: y, width: 5 * u, height: g.headTall), radius: 2.5 * u)
    }
    let head = ScrubHeadMark(shape: shape, color: opts.passedColor, alpha: fade(hx),
                             glowBlur: opts.headGlow ? 14 * u : nil)

    return ScrubFrame(flash: flash, dipAlpha: dipAlpha, band: band, track: track, ticks: ticks, head: head)
}

extension ScrubDrawing {
    /// This drawing's frame at `t` — what the app's painter strokes.
    public func frame(at t: Double, _ frame: FrameBox) -> ScrubFrame? {
        scrubFrame(plan, options, pictures, t, frame)
    }
}
